// @nactjs/core/nact-constants/HttpStatusMessages.const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTTP_STATUS_MESSAGES = void 0;
const HTTP_STATUS_MESSAGES = {
    Continue: "Continue",
    SwitchingProtocols: "Switching Protocols",
    Processing: "Processing",
    EarlyHints: "Early Hints",
    OK: "OK",
    Created: "Created",
    Accepted: "Accepted",
    NonAuthoritativeInformation: "Non-Authoritative Information",
    NoContent: "No Content",
    ResetContent: "Reset Content",
    PartialContent: "Partial Content",
    MultiStatus: "Multi-Status",
    AlreadyReported: "Already Reported",
    IMUsed: "IM Used",
    MultipleChoices: "Multiple Choices",
    MovedTemporarily: "Found, 302 Moved Temporarily",
    SeeOther: "See Other",
    NotModified: "Not Modified",
    UseProxy: "Use Proxy",
    TemporaryRedirect: "Temporary Redirect",
    BadRequest: "Bad Request",
    Unauthorized: "Unauthorized",
    PaymentRequired: "Payment Required",
    Forbidden: "Forbidden",
    NotFound: "Not Found",
    MethodNotAllowed: "Method Not Allowed",
    NotAcceptable: "Not Acceptable",
    ProxyAuthenticationRequired: "Proxy Authentication Required",
    RequestTimeout: "Request Timeout",
    Conflict: "Conflict",
    Gone: "Gone",
    LengthRequired: "Length Required",
    PreconditionFailed: "Precondition Failed",
    PayloadTooLarge: "Payload Too Large",
    URITooLong: "URI Too Long",
    UnsupportedMediaType: "Unsupported Media Type",
    RangeNotSatisfiable: "Range Not Satisfiable",
    ExpectationFailed: "Expectation Failed",
    ImAteapot: "I’m a teapot ",
    AuthenticationTimeout: "Authentication Timeout",
    MisdirectedRequest: "Misdirected Request",
    UnprocessableEntity: "Unprocessable Entity",
    Locked: "Locked",
    FailedDependency: "Failed Dependency",
    TooEarly: "Too Early",
    UpgradeRequired: "Upgrade Required",
    PreconditionRequired: "Precondition Required",
    TooManyRequests: "Too Many Requests",
    RequestHeaderFieldsTooLarge: "Request Header Fields Too Large",
    RequestedHostUnavailable: "Requested host unavailable",
    RetryWith: "Retry With",
    UnavailableForLegalReasons: "Unavailable For Legal Reasons",
    ClientClosedRequest: "Client Closed Request",
    InternalServerError: "Internal Server Error",
    NotImplemented: "Not Implemented",
    BadGateway: "Bad Gateway",
    ServiceUnavailable: "Service Unavailable",
    GatewayTimeout: "Gateway Timeout",
    HTTPVersionNotSupported: "HTTP Version Not Supported",
    VariantAlsoNegotiates: "Variant Also Negotiates",
    InsufficientStorage: "Insufficient Storage",
    LoopDetected: "Loop Detected",
    ResourceLimitReached: "Resource Limit Reached",
    BandwidthLimitExceeded: "Bandwidth Limit Exceeded",
    NotExtended: "Not Extended",
    NetworkAuthenticationRequired: "Network Authentication Required",
    UnknownError: "Unknown Error",
    WebServerIsDown: "Web Server Is Down",
    ConnectionTimedOut: "Connection Timed Out",
    OriginIsUnreachable: "Origin Is Unreachable",
    ATimeoutOccurred: "A Timeout Occurred",
    SSLHandshakeFailed: "SSL Handshake Failed",
    InvalidSSLCertificate: "Invalid SSL Certificate",
};
exports.HTTP_STATUS_MESSAGES = HTTP_STATUS_MESSAGES;
