// @nactjs/core/nact-constants/router.const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HANDLER__ALLOWED__EXPECTIONS = exports.ROUTE__CONFIG = exports.ROUTE__PATHS = exports.ROUTE__PARAMS = exports.ROUTE__PARAMETER__METADATA = exports.ROUTE__PATH = exports.ROUTE__METHOD__OPTIONS = exports.ARG_TO_CALL_DESCRIPTOR_OPTIONS = exports.CONTROLLER__CONFIG = exports.CONTROLLER__WATERMARK = exports.CONTROLLER_ROUTER__NAME = void 0;
const CONTROLLER_ROUTER__NAME = "nact__controller__router__name";
exports.CONTROLLER_ROUTER__NAME = CONTROLLER_ROUTER__NAME;
const CONTROLLER__WATERMARK = "nact__contoroller__watermark";
exports.CONTROLLER__WATERMARK = CONTROLLER__WATERMARK;
const CONTROLLER__CONFIG = "nact__controller__config";
exports.CONTROLLER__CONFIG = CONTROLLER__CONFIG;
const ARG_TO_CALL_DESCRIPTOR_OPTIONS = "controller_descriptor__option__G2lMnA_F4bfK";
exports.ARG_TO_CALL_DESCRIPTOR_OPTIONS = ARG_TO_CALL_DESCRIPTOR_OPTIONS;
const ROUTE__PATHS = "nact__route__metadata__options";
exports.ROUTE__PATHS = ROUTE__PATHS;
const ROUTE__CONFIG = "nact__route__config";
exports.ROUTE__CONFIG = ROUTE__CONFIG;
const HANDLER__ALLOWED__EXPECTIONS = "nact__handler__expections";
exports.HANDLER__ALLOWED__EXPECTIONS = HANDLER__ALLOWED__EXPECTIONS;
const ROUTE__METHOD__OPTIONS = "route__method__option";
exports.ROUTE__METHOD__OPTIONS = ROUTE__METHOD__OPTIONS;
const ROUTE__PATH = "route__path__option";
exports.ROUTE__PATH = ROUTE__PATH;
const ROUTE__PARAMETER__METADATA = "route__parameter__metadata";
exports.ROUTE__PARAMETER__METADATA = ROUTE__PARAMETER__METADATA;
const ROUTE__PARAMS = "route__http__params";
exports.ROUTE__PARAMS = ROUTE__PARAMS;
