// @nactjs/core/routing/router-class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NactRouter = void 0;
const expections_1 = require("../expections");
class NactRouter {
    child;
    absolute;
    expectionHandler;
    instance;
    constructor(controller, app) {
        this.instance = controller;
        this.absolute = [];
        this.child = {};
        this.expectionHandler = new expections_1.ControllerExpectionsHandler(app, controller);
    }
    getInstance() {
        return this.instance;
    }
    getControllerHandler() {
        return this.expectionHandler;
    }
    addRoute(route) {
        const routerName = route.path + "#" + route.method;
        if (!this.child[routerName]) {
            this.child[routerName] = route;
            if (route.absolute) {
                this.absolute.push(routerName);
            }
        }
    }
    getChild(path, method) {
        if (path) {
            const name = method ? path + "#" + method : path;
            return this.child[name];
        }
        return this.child;
    }
    getAbsolute() {
        return this.absolute;
    }
    hasAbsolute(path, method) {
        const name = method ? path + "#" + method : path;
        return this.absolute.includes(name);
    }
}
exports.NactRouter = NactRouter;
