// @nactjs/core/shared/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeSlashes = exports.isString = exports.isUppercase = exports.isUndefined = exports.isNull = exports.isDefined = exports.isObject = exports.isClassInstance = exports.isInitializedClass = void 0;
function isInitializedClass(object) {
    return object && isClassInstance(object) && typeof object === "object";
}
exports.isInitializedClass = isInitializedClass;
function isClassInstance(object) {
    return (object?.prototype?.constructor?.toString().substring(0, 5) === "class" ||
        object?.constructor?.toString().substring(0, 5) === "class");
}
exports.isClassInstance = isClassInstance;
function isUndefined(object) {
    return object === undefined;
}
exports.isUndefined = isUndefined;
function isNull(object) {
    return object === null;
}
exports.isNull = isNull;
function isDefined(object) {
    return !isUndefined(object) && !isNull(object);
}
exports.isDefined = isDefined;
function isObject(object) {
    return isDefined(object) && typeof object === "object";
}
exports.isObject = isObject;
const isUppercase = (value) => {
    return value[0] === value[0].toUpperCase();
};
exports.isUppercase = isUppercase;
const isString = (value) => {
    return typeof value === "string";
};
exports.isString = isString;
const removeSlashes = (string) => {
    if (string !== "/") {
        if (string[0] === "/")
            string = string.slice(1);
        if (string[string.length - 1] === "/")
            string = string.slice(0, string.length - 1);
    }
    return string;
};
exports.removeSlashes = removeSlashes;
